src/main/java/seedu/address/model/lesson/TimeRange.java

```diff
@@ -0,0 +1,83 @@
+package seedu.address.model.lesson;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+import java.util.Objects;
+
+/**
+ * Represents a Lesson's TimeRange in the address book.
+ * Guarantees: immutable; is valid as declared in {@link #isValidTimeRange(Time, Time)}
+ */
+public class TimeRange implements Comparable<TimeRange> {
+    public static final String MESSAGE_CONSTRAINTS = "End time cannot be earlier than start time.";
+
+    private final Time start;
+    private final Time end;
+
+    /**
+     * Construst a TimeRange object.
+     *
+     * @param start Start of the range.
+     * @param end End of the range.
+     */
+    public TimeRange(Time start, Time end) {
+        requireNonNull(start);
+        requireNonNull(end);
+        this.start = start;
+        this.end = end;
+        checkArgument(isValidTimeRange(start, end), MESSAGE_CONSTRAINTS);
+    }
+
+    /**
+     * Checks if the start is earlier than the end.
+     *
+     * @param start Start of the range.
+     * @param end End of the range.
+     * @return True if start is earlier than or same time as end.
+     */
+    public static boolean isValidTimeRange(Time start, Time end) {
+        return end.compareTo(start) >= 0;
+    }
+
+    /**
+     * Checks if the given time is within the time range.
+     *
+     * @param time The time to be tested.
+     */
+    public boolean isClashing(Time time) {
+        return start.compareTo(time) <= 0 // same or later than start time
+            && end.compareTo(time) > 0; // earlier than end time
+    }
+
+    @Override
+    public String toString() {
+        return start.toString() + " to " + end.toString();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+            || (other instanceof TimeRange) // instanceof handles nulls
+            && start.equals(((TimeRange) other).start)
+            && end.equals(((TimeRange) other).end); // state check
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(start, end);
+    }
+
+    /**
+     * Compares this TimeRange object with the other TimeRange object.
+     *
+     * @param other The TimeRange object to compare with.
+     * @return 1, if this is later than other;0 if clashing; -1 if this is earlier.
+     */
+    @Override
+    public int compareTo(TimeRange other) {
+        boolean isStartClashing = isClashing(other.start);
+        boolean isEndClashing = isClashing(other.end);
+        return isStartClashing || isEndClashing ? 0 : start.compareTo(other.start);
+    }
```

Comment:
what is this for? Should compareTo here be just return comparing of start times? So in `Lesson.compareTo` just call `Date.compareTo` and `TimeRange.compareTo`

Reply:
The starting time might be the same but the ending time may be different since it is a range, so I think cannot just compare start time. 
We are only comparing starting time of lessons if their date is the same and we can be assured that the range does not clashed (since it is handled by the time range validation logic) so we don't need to look at the ending time. Does this make sense?

Reply:
In a sense, this function means time ranges are only worth comparing if they do not clash with each other.
Equality is handled by equals method.
But in the case of TreeSet, since it uses compareTo instead of equals, I'm not sure if this will affect but the TreeSet holds Lesson where the compareTo method in Lesson does not depend on TimeRange's compareTo

Reply:
ok I think I get what you mean I'm thinking it can work this way, let me know if it makes sense.
1. `TimeRange.isClashing` take in a `TimeRange` instead of `Time` and it will check if the range overlaps. Then you can use `TimeRange.isClashing`  in `LessonAddCommand` instead of `TimeRange.compareTo`.
2. `TimeRange.compareTo` can retain the logic, if `isClashing` return 0, otherwise compare `start`
3. in `Lesson`, we can just use `TimeRange.compareTo`. If lesson has clashing timing, then it wont be accepted. Otherwise it will order by start time.

Reply:
Oh yes, I will make the changes